mpa/modules/hooks/fp16_sam_optimizer_hook.py

```diff
@@ -0,0 +1,99 @@
+# Copyright (C) 2022 Intel Corporation
+# SPDX-License-Identifier: Apache-2.0
+#
+
+import torch
+from mmcv.runner import HOOKS, Fp16OptimizerHook
+
+
+@HOOKS.register_module()
+class Fp16SAMOptimizerHook(Fp16OptimizerHook):
+    '''Sharpness-aware Minimization optimizer hook
+
+    Implemented as OptimizerHook for MMCV Runners
+    - Paper ref: https://arxiv.org/abs/2010.01412
+    - code ref: https://github.com/davda54/sam
+    '''
+
+    def __init__(self, rho=0.05, start_epoch=1, **kwargs):
+        print(f"rho {rho}, st_epoch {start_epoch}, {kwargs}")
+        super().__init__(**kwargs)
+        self.rho = rho
+        self.start_epoch = start_epoch
+        if rho < 0.0:
+            raise ValueError('rho should be greater than 0 for SAM optimizer')
+
+    def after_train_iter(self, runner):
+        '''Perform SAM optimization
+        0. compute current loss (DONE IN model.train_step())
+        1. compute current gradient
+        2. move param to the approximate local maximum: w + e(w) = w + rho*norm_grad
+        3. compute maximum loss
+        4. compute SAM gradient on maximum loss
+        5. restore parram to original param
+        6. update param using SAM gradient
+
+        Assuming model.current_batch had been set in model.train_step()
+        '''
+        current_batch = self._get_current_batch(runner.model)
+        if current_batch is None or runner.epoch+1 < self.start_epoch:
+            # Apply original parameter update
+            return super().after_train_iter(runner)
+
+        # Current gradient
+        runner.optimizer.zero_grad()
+        curr_loss = runner.outputs['loss']
+        curr_loss.backward()
+
+        # Move to local maximum
+        param2move = {}
+        with torch.no_grad():
+            scale = self.rho / (self._grad_norm(runner.optimizer) + 1e-12)
+            for param_group in runner.optimizer.param_groups:
+                for param in param_group['params']:
+                    if param.grad is None:
+                        continue
+                    e_w = param.grad * scale.to(param)
+                    param.add_(e_w)  # Climb toward gradient (increasing loss)
+                    param2move[param] = e_w  # Saving for param restore
+
+        # SAM gradient
+        runner.optimizer.zero_grad()
+        max_outputs = runner.model.train_step(current_batch, runner.optimizer)  # forward() on maxima param
+        max_loss = max_outputs['loss']
+        self.loss_scaler.scale(max_loss).backward()
```

Comment:
Are you talking about SGD gradient in OptimizerHook? loss_scaler is needed for mixed precision. It prevents gradient from becoming zero in fp16 by multiplying by scale factor to loss. And weight gradients should be unscaled before updating optimizer.

Reply:
Yes, I agree but I'm wondering `curr_loss.backward()` is right code for mixed precision. Even though I'm not sure for all details about FP16, I think that may be `self.loss_scaler.scale(curr_loss).backward()`

Reply:
I was concerned about that, too. If we want to modify the parameter before the weight update, it should be unscaled to maintain the magnitude. But unscale_ must be called once per optimizer step. But the first backward in the sam optimizer is not for parameter update, so unscale function will be duplicated causing RuntimeError. The first backward in SAM is for moving parameter to local maximum and getting larger loss for the next bp, not performing the actual parameter updating. So I thought that zero gradient would not be matter to the first backward because it will be dealt by scaling in the next backward if it happens. If you have an other opinion, let me know.